src/main/java/com/android/volley/toolbox/DiskBasedAsyncCacheFallback.java

```diff
@@ -0,0 +1,152 @@
+package com.android.volley.toolbox;
+
+import android.text.TextUtils;
+import androidx.annotation.VisibleForTesting;
+import com.android.volley.AsyncCache;
+import com.android.volley.VolleyLog;
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * AsyncCache implementation that uses Java's standard IO libraries to perform disc reads and writes
+ * This should be used as a fallback for devices with an API level below 26, as the operations are
+ * synchronous.
+ */
+public class DiskBasedAsyncCacheFallback extends AsyncCache {
+
+    /** Map of the Key, CacheHeader pairs */
+    private final Map<String, CacheHeader> mEntries = new LinkedHashMap<>(16, .75f, true);
+
+    /** Total amount of space currently used by the cache in bytes. */
+    private long mTotalSize = 0;
+
+    /** The supplier for the root directory to use for the cache. */
+    private final DiskBasedCacheUtility.FileSupplier mRootDirectorySupplier;
+
+    /** The maximum size of the cache in bytes. */
+    private final int mMaxCacheSizeInBytes;
+
+    /** Default maximum disk usage in bytes. */
+    private static final int DEFAULT_DISK_USAGE_BYTES = 5 * 1024 * 1024;
+
+    /** High water mark percentage for the cache */
+    @VisibleForTesting static final float HYSTERESIS_FACTOR = 0.9f;
+
+    /**
+     * Constructs an instance of the DiskBasedAsyncCacheFallback at the specified directory.
+     *
+     * @param rootDirectory The root directory of the cache.
+     * @param maxCacheSizeInBytes The maximum size of the cache in bytes. Note that the cache may
+     *     briefly exceed this size on disk when writing a new entry that pushes it over the limit
+     *     until the ensuing pruning completes.
+     */
+    public DiskBasedAsyncCacheFallback(final File rootDirectory, int maxCacheSizeInBytes) {
+        mRootDirectorySupplier =
+                new DiskBasedCacheUtility.FileSupplier() {
+                    @Override
+                    public File get() {
+                        return rootDirectory;
+                    }
+                };
+        mMaxCacheSizeInBytes = maxCacheSizeInBytes;
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
+    }
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public synchronized void get(String key, OnGetCompleteCallback callback) {
```

Comment:
This is really similar to the existing DiskBasedCache, except instead of returning null or entry.toCacheEntry, you call callback.onGetComplete(null/entry.toCacheEntry)

I'm wondering if it would be possible to use the existing DiskBasedCache in here and just call #get, then do callback.onGetComplete with whatever it returns. There may be a couple ways of approaching this, so happy to discuss more offline too.

I think it would generally simplify this class if you can leverage the existing implementation.

Reply:
+1 - for the fallback, we should try to share as much code with possible as the existing implementation rather than duplicating it.

In fact, though, I wonder if we really want to have a second class at all. The reality is that it's confusing to have a class extend AsyncCache but not actually be Async - how should the user of this object determine whether the cache supports true Async operation (and thus doesn't need to be called on a blocking thread) or not (and thus does)?

The class which uses this will ultimately need to make a decision based on the capabilities of the cache to either use the fully async API or else send the task to a blocking thread pool. But in the latter case, there's actually no reason to use the async API methods, since they're blocking anyway.

This might become clearer when we actually implement the async cache handling step, but my instinct here is that we actually don't need this class, and should instead have the dispatcher accept either a vanilla Cache (which defaults to DiskBasedCache) or an AsyncCache (which defaults to DiskBasedAsyncCache, where supported). If the cache is an instance of AsyncCache, it knows it can safely use the async methods. If not, it should use the fallback and schedule the read for a blocking background thread.

Reply:
I changed it so the fallback now doesn't repeat the code. But yeah, that makes sense.

Reply:
That's certainly better, but then why would we need it at all per my comment? Would it make sense to hold off on adding it unless/until we see how it would be helpful/needed, or do you see a reason that the next level up would make use of this, given that it will need to know how to dispatch the cache operations?

Reply:
oh okay, I understand what you're saying now. I just deleted it in the last commit.